crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3983,6 +4017,116 @@ impl<'db> TypeInferenceBuilder<'db> {
             }
         }
     }
+
+    fn infer_subscript_type_expression(
+        &mut self,
+        subscript: &ruff_python_ast::ExprSubscript,
+        value_ty: Type<'db>,
+    ) -> Type<'db> {
+        let ast::ExprSubscript {
+            range: _,
+            value: _,
+            slice,
+            ctx: _,
+        } = subscript;
+
+        match value_ty {
+            Type::Instance(InstanceType {
+                class: _,
+                known: Some(known_instance),
+            }) => self.infer_parameterized_known_instance_type_expression(known_instance, slice),
+            _ => {
+                self.infer_type_expression(slice);
+                Type::Todo
+            }
+        }
+    }
+
+    fn infer_parameterized_known_instance_type_expression(
+        &mut self,
+        known_instance: KnownInstance,
+        parameters: &ast::Expr,
+    ) -> Type<'db> {
+        match known_instance {
+            KnownInstance::Literal => {
+                let Some(slice_ty) = self.infer_literal_parameter_type(parameters) else {
+                    self.diagnostics.add(
+                        parameters.into(),
+                        "invalid-literal-parameter",
+                        format_args!(""),
+                    );
+                    return Type::Unknown;
+                };
+
+                match slice_ty {
+                    Type::Tuple(tuple) => {
+                        let elts = tuple.elements(self.db);
+                        UnionType::from_elements(self.db, elts)
+                    }
+                    ty => ty,
+                }
+            }
+        }
+    }
+
+    fn infer_literal_parameter_type(&mut self, parameters: &ast::Expr) -> Option<Type<'db>> {
+        Some(match parameters {
+            subscript @ ruff_python_ast::Expr::Subscript(_) => {
+                let inner_ty = self.infer_type_expression(subscript);
+                if !matches!(
+                    inner_ty,
+                    Type::IntLiteral(_)
+                        | Type::BooleanLiteral(_)
+                        | Type::BytesLiteral(_)
+                        | Type::StringLiteral(_)
+                        | Type::Union(_)
+                ) {
+                    return None;
+                }
+                inner_ty
+            }
+            ruff_python_ast::Expr::Tuple(t) if !t.parenthesized => {
```

Comment:
Are parenthesized tuples explicitly disallowed by the spec? That would surprise me if so, since `Literal[1, 2]` and `Literal[(1, 2)]` have the same AST from Python's point of view.

If they're not explicitly disallowed, I'd prefer to allow them

Reply:
So, this section of the spec here is clear that `Literal[(1, 2)]` should not mean the same thing as `tuple[Literal[1], Literal[1]]`. A big part of the reason for this is that type checkers like mypy who use Python's AST wouldn't be able to tell the difference between `Literal[(1, 2)]` and `Literal[1, 2]`: the two expressions have the same AST in Python.

The spec here does not explicitly say whether `Literal[(1, 2)]` should mean the same thing as `Literal[1, 2]` or not. However, I think it would be unreasonable to say that they should _not_ mean the same thing. As I just mentioned, type checkers like mypy that use Python's AST can't easily tell the difference, so it would be very hard for them to enforce that rule if the spec mandated it. Pyright [seems to enforce this rule anyway](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2AUNQCYnBTAAU5AnAFyHFlUDaLAIwAaKACYAlAF1JnalEVQQJAG4kqAfXgISbdpOpA), but I think that's a bit silly personally.

Curious for @carljm's thoughts here.

Reply:
By my reading, the spec is quite clear that `Literal[(1, "foo", "bar")]` is a disallowed form. It's specifically listed under forms that are "intentionally disallowed by design." If the spec meant to say "this form is valid, with this meaning", then it could and would say that, instead of saying that it's an invalid form. It even says `Literal[(1, 2, 3)]` "could easily be confused with" `Literal[1, 2, 3]`, which is a very weird thing to say if the intended meaning is that those should be equivalent.

An argument could be made that the spec should be changed to make `Literal[(1, 2, 3)]` equivalent to `Literal[1, 2, 3]`, because the Python typing spec shouldn't distinguish forms that can't be distinguished in the CPython AST. I have mixed feelings about that argument, but I don't care too much either way. But we _can_ distinguish these forms, and it's a lot easier to disallow something now and allow it in the future than the other way around, so I think we should follow the spec and disallow it.

Reply:
It seems especially unlikely to me that this really was the intended meaning of the spec, given that the PEP that introduced `Literal` (which was copied over to the spec verbatim) was written by a team of mypy developers. It's very hard for me to believe that they really wanted to specify something here that they knew they wouldn't be able to enforce in their reference implementation of the feature; it seems to me much more plausible that they meant only to say that parameterising `Literal` with a tuple literal wouldn't imply a tuple type.

But I don't feel too strongly, and I agree that it's easier to move from disallowed->allowed than vice versa. So let's leave things as @Glyphack has them for now 👍